README.md
# 抖店 Python SDK

[![PyPI version](https://badge.fury.io/py/doudian.svg)](https://badge.fury.io/py/doudian)
[![Download count](https://img.shields.io/pypi/dm/doudian)](https://img.shields.io/pypi/dm/doudian)

## 介绍

抖店 Python SDK。

欢迎抖店相关开发者扫码进 QQ 群(群号：799137292)讨论：

![image](qq.png)

## 适用对象

**抖店 Python SDK**同时支持自用型应用和工具型应用，具体说明详见[官网](https://op.jinritemai.com/docs/guide-docs/127/17)。

## 特性

1. 自动维护 access_token 的更新；
2. 支持本地缓存 access_token；
3. 业务参数自动排序，无需预处理；
4. 消息推送自动验证，自动解析；
5. 支持沙箱环境[测试店铺](https://op.jinritemai.com/docs/guide-docs/129/209)。

## 源码

[github](https://github.com/minibear2021/doudian)

[gitee](https://gitee.com/minibear2021/doudian)

## 安装

```
$ pip install doudian
```

## 使用方法

### 准备

参考抖店官方文档准备好 app key、app secret、shop id 等，具体见官方开发文档。

### 一个最小的后端

[examples.py](examples.py) 演示了一个带有[会员订单列表查询](https://op.jinritemai.com/docs/api-docs/13/366)接口和[消息推送处理](https://op.jinritemai.com/docs/guide-docs/153/99)接口的后端。
首先，修改 **examplys.py** 里以下几项配置参数：

```python
# App类型，SELF=自用型应用, TOOL=工具型应用
APP_TYPE = AppType.SELF

# 应用key，长度19位数字字符串
APP_KEY = '3409409348479354011'

# 应用密钥 字符串
APP_SECRET = '2ad2355c-01d0-11f8-91dc-05a8cd1054b1'

# 店铺ID，自用型应用必传。
SHOP_ID = '323423'

# token缓存文件
TOKEN_FILE = './323423.token'

# 日志记录器，记录web请求和回调细节
logging.basicConfig(filename=os.path.join(os.getcwd(), 'demo.log'), level=logging.DEBUG, filemode='a', format='%(asctime)s - %(process)s - %(levelname)s: %(message)s')
LOGGER = logging.getLogger("demo")

# 代理设置，None或者{"https": "http://10.10.1.10:1080"}，详细格式参见https://docs.python-requests.org/zh_CN/latest/user/advanced.html
PROXY = None

# 沙箱模式测试店铺
TEST_MODE = False
```

接下来初始化 DouDian 实例并配置一个合适的接口：

```python
# 初始化
doudian = DouDian(
    app_key=APP_KEY,
    app_secret=APP_SECRET,
    app_type=APP_TYPE,
    shop_id=SHOP_ID,
    token_file=TOKEN_FILE,
    logger=LOGGER,
    proxy=PROXY,
    test_mode=TEST_MODE
)

app = Flask(__name__)

@app.route('/brandList')
def brand_list():
    # 以获取店铺的已授权品牌列表为例，参照官方文档，将path和params拼凑好传入request接口，调用成功后即可获取店铺的已授权品牌列表数据。
    # https://op.jinritemai.com/docs/api-docs/13/54
    path = '/shop/brandList'
    params = {}
    result = doudian.request(path=path, params=params)
    return jsonify({'result': result if result else ''})

```

检查一下参数无误，现在就可以用 python 解释器来运行：

```shell
$ python examples.py
 * Serving Flask app "examples" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

现在访问 http://127.0.0.1:5000/orderList ，如果一切正常，你会看到类似下面一串 json 字符串：

```python
{
    "result":
    {
        "code":10000,
        "data":[],
        "err_no":0,
        "log_id":"20211203091350010194059098259B567F",
        "message":"success",
        "msg":"success",
        "sub_code":"",
        "sub_msg":""
    }
}
```

到这一步一个最小化的后端就完成了。

**以上步骤如果不能正确执行，务必仔细检查各项初始化参数，必要的情况下，登录抖店后台，将所有参数重置。**

下面配置一个消息推送接口，以便接收处理抖店提供的[消息推送服务](https://op.jinritemai.com/docs/guide-docs/153/99)。

```python
@app.route('/notify', methods=['POST'])
def notify():
    result = doudian.callback(request.headers, request.data)
    if result:
        tag = result.get('tag')
        if tag == '0':  # 抖店推送服务验证消息，需立即返回success
            return jsonify({'code': 0, 'msg': 'success'})
        if tag == '100':  # 订单创建消息，更多消息类型查阅官方文档。
            # TODO: 根据推送的消息参数进行必要的业务处理，5秒内返回success
            return jsonify({'code': 0, 'msg': 'success'})
        return jsonify({'code': 0, 'msg': 'success'})
    else:
        return jsonify({'code': 40041, 'message': '解析推送数据失败'})
```

消息推送接口上线后，将 url 地址配置到抖店后台，抖店服务器将自动推送一条 tag 值为'0'的验证消息，验证通过后方能启用消息推送服务。

## 通用接口

### request(path: str, params: dict)

所有抖店的 API 调用都通过**doudian.request()**接口进行。此接口需要传入 path 和 params 两个参数。
参数值参照官方文档组织，形式类似下面的代码，其中 params 只需要传入请求参数或业务参数，公共参数无需传入，SDK 内部会自行处理：
![image](param.png)

```python
path = '/member/searchList'
params = {}
params.update({'app_id':1})
params.update({'start_time': '2021/10/12 00:00:00'})
params.update({'end_time': '2021/10/12 00:00:00'})
params.update({'page':1})
params.update({'size':50})
result = doudian.request(path=path, params=params)
```

### init_token(code: str)

用于初始化或重置 access token，通常情况下无需手工调用，仅当以下场景时才需要：

1. 工具型（AppType.TOOL）应用初始化 DouDian 的时候未传入 code，可在获取到 code 时调用；
2. 工具型（AppType.TOOL）或自用型（AppType.SELF）应用需要强制重置 access token 时调用;
3. 捕获到 TokenError 异常时调用。

### callback(headers: dict, body: bytes)

抖店消息推送解析验证接口，将收到的推送 headers 和 body 原样传入，接口内自动解析并验证。

```python
result = doudian.callback(request.headers, request.data)
```

### build_auth_url(service_id: str, state: str)

工具型应用授权 url 构造接口，生成的授权 url 发送给商户完成授权验证以获取 code。获取到的 code 可以调用 init_token(code)接口进行 access token 的初始化或重置。

```python
service_id = 'demo service id'
state = 'demo state'
result = doudian.build_auth_url(service_id=service_id, state=state)
```

以上接口的调用示例可以参考[examples.py](examples.py)。

### 接口函数参数

参数类型对照参考下表：

| 抖店 API 官方文档声明 | **doudian python sdk** |
| --------------------- | ---------------------- |
| String                | str                    |
| Int64, Number         | int                    |
| object, Json, Struct  | dict: {}               |
| List, JsonArray       | list: []               |
| Bool                  | bool: True, False      |

### 接口函数返回值

每个接口均返回 None 或 json。

## 常见问题

### 消息推送解析失败处理

开发者遇到的难点之一就是消息推送解析失败的问题，由于众多的 python web 框架对回调消息的处理不完全一致，如果出现回调验证失败，请务必确认传入的 headers 和 body 的值和类型。
通常框架传过来的 headers 类型是 dict，而 body 类型是 bytes。使用以下方法可直接获取到解密后的实际内容。

#### flask 框架

直接传入 request.headers 和 request.data 即可。

```python
result = doudian.callback(headers=request.headers, body=request.data)
```

#### django 框架

由于 django 框架特殊性，会将 headers 做一定的预处理，可以参考以下方式调用。

```python
headers = {}
headers.update({'app-id': request.META.get('HTTP_APP_ID')})
headers.update({'event-sign': request.META.get('HTTP_EVENT_SIGN')})
result = doudian.callback(headers=headers, body=request.body)
```

#### tornado 框架

直接传入 request.headers 和 request.body 即可。

```python
result = doudian.callback(headers=request.headers, body=request.body)
```

#### 其他框架

参考以上处理方法，大原则就是保证传给 callback 的参数值和收到的值一致，不要转换为 dict，也不要转换为 string。


CHANGELOG.md
# Changelog

## [0.6] - 2022-01-25

### Fixed

- 规避param_json中带#的情况

## [0.5] - 2021-12-21

### Fixed

- 加载 token 缓存文件的 bug

## [0.4] - 2021-12-03

### Changed

- 简化 doudian.request() 接口参数

## [0.3] - 2021-12-02

### Fixed

- 请求参数缺失
- 请求 url 错误

## [0.2] - 2021-12-02

### Added

- 支持沙箱环境测试店铺
- 明确异常类型

## [0.1] - 2021-12-01

### Added

- 初始版本


doudian/core.py
# -*- coding: utf-8 -*-
import json
import time
from datetime import datetime
from logging import Logger

import requests
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.hashes import MD5, SHA256, Hash
from cryptography.hazmat.primitives.hmac import HMAC

from .exception import CodeError, ShopIdError, TokenError
from .type import AppType


class DouDian():
    def __init__(self, app_key: str, app_secret: str, app_type: AppType = AppType.SELF, code=None, shop_id: str = None, token_file: str = None, logger: Logger = None, proxy: str = None, test_mode=False):
        """初始化DouDian实例，自用型应用传入shop_id用于初始化access token，工具型应用传入code换取access token（如初始化时未传入，可以在访问抖店API之前调用init_token(code)进行token的初始化。
        """
        self._app_key = app_key
        self._app_secret = app_secret
        self._app_type = app_type
        self._token_file = token_file
        self._logger = logger
        if self._app_type == AppType.SELF and not shop_id:
            if self._logger:
                self._logger.exception('shop_id is not assigned.')
            raise ShopIdError('shop_id is not assigned.')
        self._shop_id = shop_id
        self._proxy = proxy
        self._test_mode = test_mode
        if self._test_mode:
            self._gate_way = 'https://openapi-sandbox.jinritemai.com'
        else:
            self._gate_way = 'https://openapi-fxg.jinritemai.com'
        self._version = 2
        self._token = None
        self._sign_method = 'hmac-sha256'
        if self._token_file:
            try:
                with open(self._token_file) as f:
                    self._token = json.load(f)
            except Exception as e:
                if self._logger:
                    self._logger.exception('{}'.format(e))
        if self._app_type == AppType.SELF:
            self.init_token()
        elif self._app_type == AppType.TOOL and code:
            self.init_token(code)

    def _sign(self, method: str, param_json: str, timestamp: str) -> str:
        param_pattern = 'app_key{}method{}param_json{}timestamp{}v{}'.format(self._app_key, method, param_json, timestamp, self._version)
        sign_pattern = '{}{}{}'.format(self._app_secret, param_pattern, self._app_secret)
        return self._hash_hmac(sign_pattern)

    def _hash_hmac(self, pattern: str) -> str:
        try:
            hmac = HMAC(key=self._app_secret.encode('UTF-8'), algorithm=SHA256(), backend=default_backend())
            hmac.update(pattern.encode('UTF-8'))
            signature = hmac.finalize()
            return signature.hex()
        except Exception as e:
            if self._logger:
                self._logger.exception('{}'.format(e))
            return None

    def _access_token(self) -> str:
        if not self._token:
            if self._logger:
                self._logger.exception('no token info, call init_token() to initialize it.')
            raise TokenError('no token info, call init_token() to initialize it.')
        try:
            if self._token.get('expires_in') - int(time.time()) < 3000:
                self._refresh_token()
            return self._token.get('access_token')
        except Exception as e:
            if self._logger:
                self._logger.exception('{}'.format(e))
            return None

    def init_token(self, code: str = '') -> bool:
        """初始化access token
        :param code: 工具型应用从授权url回调中获取到的code，自用型应用无需传入。
        """
        if self._app_type == AppType.TOOL and not code:
            if self._logger:
                self._logger.exception('code is not assigned.')
            raise CodeError('code is not assigned.')
        path = '/token/create'
        grant_type = 'authorization_self' if self._app_type == AppType.SELF else 'authorization_code'
        params = {}
        params.update({'code': code if code else ''})
        params.update({'grant_type': grant_type})
        if self._app_type == AppType.SELF:
            if self._test_mode:
                params.update({'test_shop': '1'})
            elif self._shop_id:
                params.update({'shop_id': self._shop_id})
            else:
                if self._logger:
                    self._logger.exception('shop_id is not assigned.')
                raise ShopIdError('shop_id is not assigned.')
        result = self._request(path=path, params=params, token_request=True)
        if result and result.get('err_no') == 0 and result.get('data'):
            self._token = result.get('data')
            self._token.update({'expires_in': int(time.time()) + result.get('data').get('expires_in')})
            if self._token_file:
                with open(self._token_file, mode='w') as f:
                    f.write(json.dumps(self._token))
            return True
        return False

    def _get_refresh_token(self) -> str:
        if not self._token:
            return None
        try:
            return self._token.get('refresh_token')
        except Exception as e:
            if self._logger:
                self._logger.exception('{}'.format(e))
            return None

    def _refresh_token(self) -> None:
        path = '/token/refresh'
        refresh_token = self._get_refresh_token()
        grant_type = 'refresh_token'
        params = {}
        params.update({'grant_type': grant_type})
        params.update({'refresh_token': refresh_token})
        result = self._request(path=path, params=params, token_request=True)
        if result and result.get('err_no') == 0 and result.get('data'):
            self._token = result.get('data')
            self._token.update({'expires_in': int(time.time()) + result.get('data').get('expires_in')})
            if self._token_file:
                with open(self._token_file, mode='w') as f:
                    f.write(json.dumps(self._token))

    def _request(self, path: str, params: dict, token_request: bool = False) -> json:
        try:
            headers = {}
            headers.update({'Content-Type': 'application/json'})
            headers.update({'Accept': 'application/json'})
            headers.update({'User-Agent': 'doudian python sdk(https://github.com/minibear2021/doudian)'})
            timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            param_json = json.dumps(params, sort_keys=True, separators=(',', ':'))
            method = path[1:].replace('/', '.')
            sign = self._sign(method=method, param_json=param_json, timestamp=timestamp)
            if token_request:
                url = self._gate_way + '{}?app_key={}&method={}&param_json={}&timestamp={}&v={}&sign_method={}&sign={}'.format(
                    path, self._app_key, method, param_json, timestamp, self._version, self._sign_method, sign)
            else:
                access_token = self._access_token()
                url = self._gate_way + '{}?app_key={}&method={}&access_token={}&timestamp={}&v={}&sign_method={}&sign={}'.format(
                    path, self._app_key, method, access_token, timestamp, self._version, self._sign_method, sign)
            if self._logger:
                self._logger.debug('Request url: {}'.format(url))
                self._logger.debug('Request headers: {}'.format(headers))
                self._logger.debug('Request params: {}'.format(param_json))
            response = requests.post(url=url, data=param_json, headers=headers, proxies=self._proxy)
            if self._logger:
                self._logger.debug('Response status code: {}'.format(response.status_code))
                self._logger.debug('Response headers: {}'.format(response.headers))
                self._logger.debug('Response content: {}'.format(response.content.decode('UTF-8')))
            if response.status_code != 200:
                return None
            return json.loads(response.content)
        except Exception as e:
            if self._logger:
                self._logger.exception('{}'.format(e))
            return None

    def request(self, path: str, params: dict) -> json:
        """请求抖店API接口
        :param path: 调用的API接口地址，示例：'/material/uploadImageSync'
        :param params: 业务参数字典，示例：{'folder_id':'70031975314169695161250','url':'http://www.demo.com/demo.jpg','material_name':'demo.jpg'}
        """
        return self._request(path=path, params=params)

    def callback(self, headers: dict, body: bytes) -> json:
        """验证处理消息推送服务收到信息
        """
        data: str = body.decode('UTF-8')
        if self._logger:
            self._logger.debug('Callback Header: %s' % headers)
            self._logger.debug('Callback Body: %s' % body)
        if not data:
            return None
        if headers.get('app-id') != self._app_key:
            return None
        event_sign: str = headers.get('event-sign')
        if not event_sign:
            return None
        h = Hash(algorithm=MD5(), backend=default_backend())
        h.update('{}{}{}'.format(self._app_key, data, self._app_secret).encode('UTF-8'))
        if h.finalize().hex() != event_sign:
            return None
        return json.loads(data)

    def build_auth_url(self, service_id: str, state: str) -> str:
        """拼接授权URL，引导商家点击完成授权
        """
        if self._app_type == AppType.TOOL:
            return 'https://fuwu.jinritemai.com/authorize?service_id={}&state={}'.format(service_id, state)
        else:
            return None


examples.py
# -*- coding: utf-8 -*-
import logging
import os
from random import sample
from string import digits, ascii_letters
from time import time

from flask import Flask, jsonify, request

from doudian import DouDian, AppType

# App类型，SELF=自用型应用, TOOL=工具型应用
APP_TYPE = AppType.SELF

# 应用key，长度19位数字字符串
APP_KEY = '3409409348479354011'

# 应用密钥 字符串
APP_SECRET = '2ad2355c-01d0-11f8-91dc-05a8cd1054b1'

# 店铺ID，自用型应用必传。
SHOP_ID = '323423'

# token缓存文件
TOKEN_FILE = './323423.token'

# 日志记录器，记录API请求和推送消息细节
logging.basicConfig(filename=os.path.join(os.getcwd(), 'demo.log'), level=logging.DEBUG, filemode='a', format='%(asctime)s - %(process)s - %(levelname)s: %(message)s')
LOGGER = logging.getLogger("demo")

# 代理设置，None或者{"https": "http://10.10.1.10:1080"}，详细格式参见https://docs.python-requests.org/zh_CN/latest/user/advanced.html
PROXY = None

# 沙箱模式测试店铺
# https://op.jinritemai.com/docs/guide-docs/129/209
TEST_MODE = True

# 初始化
doudian = DouDian(
    app_key=APP_KEY,
    app_secret=APP_SECRET,
    app_type=APP_TYPE,
    shop_id=SHOP_ID,
    token_file=TOKEN_FILE,
    logger=LOGGER,
    proxy=PROXY,
    test_mode=TEST_MODE
)

app = Flask(__name__)


@app.route('/brandList')
def brand_list():
    # 以获取店铺的已授权品牌列表为例，参照官方文档，将path和params拼凑好传入request接口，调用成功后即可获取店铺的已授权品牌列表数据。
    # https://op.jinritemai.com/docs/api-docs/13/54
    path = '/shop/brandList'
    params = {}
    result = doudian.request(path=path, params=params)
    return jsonify({'result': result if result else ''})


@app.route('/orderList')
def order_list():
    # 以会员订单列表查询为例，参照官方文档，将path和params拼凑好传入request接口，调用成功后即可获取会员订单数据。
    # https://op.jinritemai.com/docs/api-docs/13/366
    path = '/member/searchList'
    params = {}
    params.update({'start_time': '2021/10/12 00:00:00'})
    params.update({'end_time': '2021/10/13 00:00:00'})
    result = doudian.request(path=path, params=params)
    return jsonify({'result': result if result else ''})


@app.route('/notify', methods=['POST'])
def notify():
    # 解析处理消息推送服务
    # https://op.jinritemai.com/docs/guide-docs/153/99
    result = doudian.callback(request.headers, request.data)
    if result:
        tag = result.get('tag')
        if tag == '0':  # 抖店推送服务验证消息，需立即返回success
            return jsonify({'code': 0, 'msg': 'success'})
        if tag == '100':  # 订单创建消息，更多消息类型查阅官方文档。
            # TODO: 根据推送的消息参数进行必要的业务处理，5秒内返回success
            return jsonify({'code': 0, 'msg': 'success'})
        return jsonify({'code': 0, 'msg': 'success'})
    else:
        return jsonify({'code': 40041, 'message': '解析推送数据失败'})


@app.route('/authUrl')
def authUrl():
    # 工具型应用生成授权url供商户进行授权操作
    # https://op.jinritemai.com/docs/guide-docs/9/22
    service_id = 'demo service id'
    state = 'demo state'
    result = doudian.build_auth_url(service_id=service_id, state=state)
    return jsonify({'result': result if result else ''})


if __name__ == '__main__':
    app.run()
